File: genesis/source/core/frame.cpp
//
// core/frame.cpp
//
// Copyright 2008-2011 Mike Austin
// All rights reserved.
//

#include "core/frame.h"
#include "core/protos/symbol.h"

using std::stringstream;

namespace impulse {

 //
 // Constructors
 //

	inline void *Frame::operator new( size_t size )
	{
		Frame* frame = (Frame *) ::operator new( size );

		if (_releasePoolStack.size() > 0)
			_releasePoolStack.back().push_back( frame );
		else
			TRACE( "No release pool available." );
		
		return frame;
	}

	inline Frame& Frame::create() { return *new Frame(); }

	inline Frame& Frame::create( Frame& proto ) { return *new Frame( proto ); }

	inline Frame::Frame()               : _protoFrame( NULL ), _publicSlots( NULL ), _referenceCount( 1 ) { }

 	inline Frame::Frame( Frame& proto ) : _protoFrame( &proto ), _publicSlots( NULL ), _referenceCount( 1 )
 	{
 		getProto().incrementReference();
 	}

	inline Frame::~Frame()
	{
		//ENTER( "Frame::~Frame()" );
		
		if (_protoFrame) getProto().decrementReference();
		if (_publicSlots) delete _publicSlots;

		//LEAVE( "Frame::~Frame()" );
	}

 //
 // Slot Access
 //

	inline Value Frame::setSlot( const Symbol symbol, const Value value )
	{
		// TODO: Why doesn't insert work the same as []?
		//getSlots().insert( make_pair( symbol.getId(), value ) );
		getSlots()[symbol.getId()] = value;

		return value;
	}

	inline Value Frame::getSlot( const Symbol symbol )
	{
		SlotMap::iterator iter = getSlots().find( symbol.getId() );
		
		if (iter != getSlots().end())
		{
			return iter->second;
		}

		return Value();
	}

	inline Value Frame::setSlot( const string name, const Value value )
	{
		return setSlot( SymbolProto::at( name ), value );
	}
	
	inline Value Frame::getSlot( const string name )
	{
		return getSlot( SymbolProto::at( name ) );
	}

 //
 // Messaging
 //

	inline Value Frame::perform( Value receiver, const Symbol selector, const Array& args, Value locals )
	{
		ENTER( "Frame::perform( receiver = " << receiver << ", selector = " << selector << " )" );
		
		Frame* frame = this;
		Value result;
		
		while (frame)
		{
			if (frame->_publicSlots)
			{
				SlotMap::iterator iter = frame->_publicSlots->find( selector.getId() );
			
				if (iter != frame->_publicSlots->end())
				{
					Value method = iter->second;
				
					result = method.apply( receiver, args, locals );
					
					LEAVE( result );
					
					return result;
				}
			}
						
			frame = frame->_protoFrame;
		}

		cerr << "Slot not found: " << receiver.inspect() << "." << selector.getName() << endl;

		LEAVE( result );

		return result;
	}

 //
 // Inspection
 //

	string Frame::inspect( const Value self ) const
	{
		stringstream stream;
		
		stream << "<frame@" << this << ">";
		
		return stream.str();
	}

	string Frame::inspect( const Value self, string name ) const
	{
		stringstream stream;
		
		stream << "<" << name << "@" << this << ">";
		
		return stream.str();
	}

 //
 // Garbage Collection
 //

	inline void Frame::incrementReference()
	{
		++_referenceCount;

		//TRACE( "\t\t\t\t\t\t\t\t+ " << inspect( *this ) << " : " << _referenceCount );
	}

	inline void Frame::decrementReference()
	{
		--_referenceCount;

		if (_referenceCount < 0) TRACE( "deleting freed object " << this );

		//if (debugGarbage) TRACE( "\t\t\t\t\t\t\t\t- " << inspect( *this ) << (_refCount == 0 ? " free" : "") );
		//TRACE( "\t\t\t\t\t\t\t\t- " << inspect( *this ) << " : " << _referenceCount <<  (_referenceCount == 0 ? " (freeing)" : "") );

		if (_referenceCount == 0)
		{
			//if (debugGarbage) TRACE( "freeing " << inspect( *this ) );

			delete this;
		}
	}

	vector< vector<Frame*> > Frame::_releasePoolStack;

 //
 // ReleasePool
 //
 
	inline Frame::ReleasePool::ReleasePool()
	{
		TRACE( "Pushing new release pool..." );
		
		Frame::_releasePoolStack.push_back( vector<Frame*>() );
	}

	inline Frame::ReleasePool::~ReleasePool()
	{
		TRACE( "Poping old release pool..." );
		
		vector<Frame*> pool = _releasePoolStack.back();
		vector<Frame*>::iterator iter = pool.begin();
		
		while (iter != pool.end())
		{
			(*iter)->decrementReference();
		
			iter = pool.erase( iter );
		}
		
		_releasePoolStack.pop_back();
	}

}

File: ruby2/backup/frame.rb
#
# frame.rb
#

class Hash

  attr :parent, true
  
  def find(symbol)
    hash = self
    
    begin
      value = hash[symbol]
      hash  = hash.parent
    end while value == nil && hash != nil
    
    if value
      return value
    #else
    #  raise "Slot not found: " + self.class.name + "." + symbol.to_s
    end

    return nil
  end

end


class Frame < Object

  attr :proto, true
  attr :slots, true
  attr :_methods, true

  def self.new(*args)
    result = super(*args)
    
    return Value.new(result)
  end

  def initialize(proto = nil)
    @proto, @slots, @_methods = proto, {}, {}
    
    if @proto != nil
      @slots.parent = @proto.slots
      @_methods.parent = @proto._methods
    else
      @slots.parent = nil
      @_methods.parent = nil
    end
  end

  def init_slots()
  end

  def set_slot(symbol, value)
    return @slots[symbol] = value
  end

  def get_slot(symbol)
    return @slots[symbol]
  end

  def find_slot(symbol)
    value = @slots.find(symbol)
    
    if value
      return value
    else
      raise "Slot not found: " + self.class.name + "." + symbol.to_s
    end
  end

  def add_method(symbol, value)
    return @_methods[symbol] = value
  end

  def get_method(symbol)
    return @_methods[symbol]
  end

  def find_method(symbol)
    return @_methods.find(symbol)
  end

  def send_(selector, receiver, args)
    method = find_method(selector)
    
    if method
      return method.frame.call_(receiver, args)
    else
      raise "Method not found: " + self.class.name + "." + selector.to_s
    end
    
    return nil
  end

  def xsend_(selector, receiver, args)
    trace "Frame::send()"

    frame = receiver

    begin
      value = frame.get_method(selector)
      frame = frame.proto
    end while value == nil && frame != nil
    
    if value
      #return value.eval_(receiver, args, nil)
      return value.frame.call_(receiver, args)
    else
      raise "Slot not found: " + self.class.name + "." + selector.to_s
    end
    
    return nil
  end

  def eval_(receiver, args, locals)
    return Value.new(self)
  end
  
end
